vllm/entrypoints/chat_utils.py

```diff
@@ -401,29 +437,42 @@ def _parse_chat_message_content_parts(
 
     has_image = False
     for part in parts:
-        part_type = part["type"]
-        if part_type == "text":
-            text = _TextParser(part)["text"]
+        if isinstance(part, str):
+            text = _TextParser(part)
             texts.append(text)
-        elif part_type == "image_url":
-            image_url = _ImageParser(part)["image_url"]
-
-            if image_url.get("detail", "auto") != "auto":
-                logger.warning(
-                    "'image_url.detail' is currently not supported and "
-                    "will be ignored.")
-
-            mm_parser.parse_image(image_url["url"])
+        elif _is_simple_image_part(part):
+            mm_parser.parse_image(part["image_url"])  # type: ignore
             has_image = True
-        elif part_type == "audio_url":
-            audio_url = _AudioParser(part)["audio_url"]
-
-            mm_parser.parse_audio(audio_url["url"])
-        elif part_type == "refusal":
-            text = _RefusalParser(part)["refusal"]
-            texts.append(text)
+        elif _is_simple_audio_part(part):
+            mm_parser.parse_audio(part["audio_url"])  # type: ignore
```

Comment:
I'm thinking about a more concise way like the following:

```python
if isinstance(part, str):
    text = _TextParser(part)
    texts.append(text)
else:
    assert isinstance(part, dict)
    part_type, part_url = parse_mm_part()
    
    if part_type == "text":
        ...
    else:
        if part_type == "image_url":
            mm_parser.parse_image(part_url)
            has_image = True
        elif part_type == "audio_url":
            ...
```

```python
def parse_mm_part(part: ChatCompletionContentPartParam):
    part_type = part.get("type", None)
    if part_type == "image_url":
        image_url = _ImageParser(part)["image_url"]
        if image_url.get("detail", "auto") != "auto":
            logger.warning(
                "'image_url.detail' is currently not supported and "
                "will be ignored.")
        url = image_url["url"]
    elif part_type == "audio_url":
        ...
    elif part_type is None:
        if "image_url" in part and isinstance(part["image_url"], str):
            part_type = "image_url"
            url = part["image_url"]
        elif ...
    else:
        ValueError(...)
    return part_type, url
```

WDYT?

Reply:
this makes more sense! let me do this